crate/operator/operations.py

```diff
@@ -156,31 +125,91 @@ async def restart_cluster(
 
-    :param coapi: An instance of the Kubernetes CustomObjects API.
     :param core: An instance of the Kubernetes Core V1 API.
     :param namespace: The Kubernetes namespace where to look up CrateDB cluster.
     :param name: The CrateDB custom resource name defining the CrateDB cluster.
-    :param total_nodes: The total number of nodes that the cluster should
-        consist of, per the CrateDB cluster spec.
+    :param old: The old resource body.
     """
-    cluster = await coapi.get_namespaced_custom_object(
-        group=API_GROUP,
-        version="v1",
-        plural=RESOURCE_CRATEDB,
-        namespace=namespace,
-        name=name,
-    )
-    password = await get_system_user_password(core, namespace, name)
-    host = await get_host(core, namespace, name)
-    conn_factory = connection_factory(host, password)
-
-    if "master" in cluster["spec"]["nodes"]:
-        await restart_statefulset(
-            core, conn_factory, namespace, name, "master", total_nodes, logger
+    pending_pods: List[Dict[str, str]] = status.get("pendingPods", [])
+    if not pending_pods:
+        pending_pods = []
+        if "master" in old["spec"]["nodes"]:
+            pending_pods.extend(
+                await get_pods_in_statefulset(core, namespace, name, "master")
+            )
+        for node_spec in old["spec"]["nodes"]["data"]:
+            pending_pods.extend(
+                await get_pods_in_statefulset(core, namespace, name, node_spec["name"])
+            )
+        patch.status["pendingPods"] = pending_pods
+
+    if not pending_pods:
+        # We're all done
+        patch.status["pendingPods"] = None  # Remove attribute from status stanca
+        return
+
+    next_pod_uid = pending_pods[0]["uid"]
+    next_pod_name = pending_pods[0]["name"]
+
+    all_pod_uids, all_pod_names = await get_pods_in_cluster(core, namespace, name)
+    if next_pod_uid in all_pod_uids:
+        # The next to-be-terminated pod still appears to be running.
+        logger.info("Terminating pod '%s'", next_pod_name)
+        # Trigger deletion of Pod.
+        # This may take a while as it tries to gracefully stop the containers
+        # of the Pod.
+        await core.delete_namespaced_pod(namespace=namespace, name=next_pod_name)
+        raise kopf.TemporaryError(
+            f"Waiting for pod {next_pod_name} ({next_pod_uid}) to be terminated.",
+            delay=BACKOFF_TIME,
         )
-    for node_spec in cluster["spec"]["nodes"]["data"]:
-        await restart_statefulset(
-            core, conn_factory, namespace, name, node_spec["name"], total_nodes, logger
+    elif next_pod_name in all_pod_names:
+        total_nodes = get_total_nodes_count(old["spec"]["nodes"])
+        # The new pod has been spawned. Only a matter of time until it's ready.
+        password = await get_system_user_password(core, namespace, name)
+        host = await get_host(core, namespace, name)
```

Comment:
Could be done in parallel:

```suggestion
        password, host = await asyncio.gather(
            get_system_user_password(core, namespace, name),
            get_host(core, namespace, name),
        )
```